clash-lib/src/Clash/Primitives/DSL.hs

```diff
@@ -352,6 +371,26 @@ outputFn _ outType _ _ texpr =
   error $ "outputFn: the expression " <> show texpr
   <> " must be an Identifier with type " <> show outType
 
+-- | Create a vector of 'TExpr's
+vec
+  :: (HasCallStack, Backend backend)
+  => [TExpr]
+  -- ^ Elements of vector
+  -> State (BlockState backend) TExpr
+  -- ^ Vector elements
+vec els@(el:_)
+  | () <- foldl checkType () els
+  = pure (TExpr (Vector (length els) (ety el)) theVec)
```

Comment:
Does `ghc` not complain about exhaustiveness here? What if the list is non-empty but not all checks pass?

Would this be clearer as `all (\x -> ety x == ety el) els` with `checkType` being removed? In the `otherwise` case you could print a list of all unique types.

Reply:
Funny, no it doesn't, but it should.

Reply:
Yes that'd be clearer, thanks :)

Reply:
Probably because there's only one _actual_ value for type `()` so you've covered all cases, it was just slightly strange to read